doc/DHCPv6_relay/DHCPv6-relay-agent-High-Level-Design.md

```diff
@@ -0,0 +1,233 @@
+# DHCPv6 Relay Agent
+
+# High Level Design Document
+
+# Table of Contents
+* [Scope](#scope)
+* [Definition](#definition)
+* [Overview](#overview)
+* [DHCPv6](#dhcpv6)
+    - [Why DHCPv6 relay agent](#why-dhcpv6-relay-agent)
+    - [DHCPv6 Relay messages](#dhcpv6-relay-messages)
+    - [DHCPv6 Packet Forwarding](#dhcpv6-packet-forwarding)
+    - [Relay Agent Behavior](#relay-agent-behavior)
+* [Requirements](#requirements)
+* [Topology](#topology)
+* [Design](#design)
+  - [CLI and Usage](#cli-and-usage)
+  - [DHCPRELAY counter](#dhcprelay-counter)
+  - [CONFIG DB schema](#config-db-schema)
+  - [YANG Model schema](#yang-model-schema)
+  - [Option 79 for client link-layer address](#option-79-for-client-link-layer-address)
+  - [Option for Dual ToR](#option-for-dual-tor)
+  - [Feature table](#feature-table)
+  - [RADV modification](#radv-modification)
+  - [CoPP manager](#copp-manager)
+  - [Source IP](#source-ip)
+* [Performance](#performance)
+* [Testing](#testing)
+
+# Scope
+
+This document describes high level design details of SONiC's DHCPv6 relay agent.
+
+# Definition
+
+DHCP: Dynamic Host Configuration Protocol
+
+DUID: DHCP Unique Identifier (Each DHCPv6 client and server has a DUID. DHCPv6 servers use it to identify clients for the selection of configuration parameters with clients. DHCPv6 clients use it to identify a server in messages where a server needs to be identified.)
+
+# Overview
+
+SONiC currently supports DHCPv4 Relay via the use of open source ISC DHCP package. However, DHCPv6 specification does not define a way to communicate client link-layer address to the DHCP server where DHCP server is not connected to the same network link as DHCP client. DHCPv6 requires all clients prepare and send a DUID as the client identifier in all DHCPv6 message exchanges. However, these methods do not provide a simple way to extract a client's link-layer address. Providing option 79 in DHCPv6 Relay-Forward messages will help carry the client link-layer address explicitly. The server needs to know the client's MAC address to allow DHCP Reservation, which provides pre-set IP address to specific client based on its physical MAC address. The DHCPv6 relay agent is able to read the source MAC address of DHCPv6 messages that it received from client, and encapsulate these messages within a DHCPv6 Relay-Forward message, inserting the client MAC address as option 79 in the Relay-Forward header sent to the server.
```

Comment:
Please add details on why we think opensource ISC package is not a right fit for this DHCP6 relay enhancements.